File: desktop/src/features/agents/conversationId.ts
import { sha256 } from "@noble/hashes/sha2.js";

const CONVERSATION_DOMAIN = new TextEncoder().encode(
  "buzz-acp-conversation-v1",
);

function decodeUuid(uuid: string): Uint8Array {
  const hex = uuid.replaceAll("-", "");
  if (!/^[0-9a-f]{32}$/i.test(hex)) {
    throw new Error(`Invalid UUID: ${uuid}`);
  }
  return Uint8Array.from({ length: 16 }, (_, index) =>
    Number.parseInt(hex.slice(index * 2, index * 2 + 2), 16),
  );
}

function formatUuid(bytes: Uint8Array): string {
  const hex = Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0"));
  return [
    hex.slice(0, 4).join(""),
    hex.slice(4, 6).join(""),
    hex.slice(6, 8).join(""),
    hex.slice(8, 10).join(""),
    hex.slice(10, 16).join(""),
  ].join("-");
}

export function deriveAgentConversationId(
  channelId: string,
  rootEventId: string,
): string {
  if (!/^[0-9a-f]{64}$/.test(rootEventId)) {
    throw new Error(`Invalid root event ID: ${rootEventId}`);
  }
  const channelBytes = decodeUuid(channelId);
  const rootBytes = new TextEncoder().encode(rootEventId);
  const input = new Uint8Array(
    CONVERSATION_DOMAIN.length + channelBytes.length + rootBytes.length,
  );
  input.set(CONVERSATION_DOMAIN);
  input.set(channelBytes, CONVERSATION_DOMAIN.length);
  input.set(rootBytes, CONVERSATION_DOMAIN.length + channelBytes.length);
  return formatUuid(sha256(input).slice(0, 16));
}

export function deriveAgentConversationIdOrNull(
  channelId: string | null | undefined,
  rootEventId: string | null | undefined,
): string | null {
  return channelId && rootEventId
    ? deriveAgentConversationId(channelId, rootEventId)
    : null;
}

File: mobile/lib/features/channels/agent_activity/conversation_id.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:pointycastle/digests/sha256.dart';

const _conversationDomain = 'buzz-acp-conversation-v1';

/// Byte-identical to desktop `deriveAgentConversationId` / Rust `deterministic_id`.
///
/// `sha256("buzz-acp-conversation-v1" + channelUuidRaw16 + rootEventIdAsciiHex)[0..16]`
/// formatted as a UUID. Channel id is raw UUID bytes; root event id is ASCII hex.
String deriveAgentConversationId(String channelId, String rootEventId) {
  if (!RegExp(r'^[0-9a-f]{64}$').hasMatch(rootEventId)) {
    throw ArgumentError.value(
      rootEventId,
      'rootEventId',
      'Invalid root event ID',
    );
  }
  final channelBytes = _decodeUuid(channelId);
  final rootBytes = utf8.encode(rootEventId);
  final domainBytes = utf8.encode(_conversationDomain);
  final input = Uint8List(
    domainBytes.length + channelBytes.length + rootBytes.length,
  );
  input.setAll(0, domainBytes);
  input.setAll(domainBytes.length, channelBytes);
  input.setAll(domainBytes.length + channelBytes.length, rootBytes);
  final digest = SHA256Digest().process(input);
  return _formatUuid(Uint8List.sublistView(digest, 0, 16));
}

/// Same as [deriveAgentConversationId], returning null on malformed input.
String? deriveAgentConversationIdOrNull(
  String? channelId,
  String? rootEventId,
) {
  if (channelId == null ||
      channelId.isEmpty ||
      rootEventId == null ||
      rootEventId.isEmpty) {
    return null;
  }
  try {
    return deriveAgentConversationId(channelId, rootEventId);
  } catch (_) {
    return null;
  }
}

/// Conversation identity for a rendered surface.
///
/// DMs use the channel id directly (Rust `id_for_event` when `is_dm`).
/// Channel threads hash `(channelId, rootEventId)`.
String? conversationIdForSurface({
  required String channelId,
  required bool isDm,
  String? rootEventId,
}) {
  if (isDm) return channelId;
  return deriveAgentConversationIdOrNull(channelId, rootEventId);
}

Uint8List _decodeUuid(String uuid) {
  final hex = uuid.replaceAll('-', '');
  if (!RegExp(r'^[0-9a-fA-F]{32}$').hasMatch(hex)) {
    throw ArgumentError.value(uuid, 'channelId', 'Invalid UUID');
  }
  return Uint8List.fromList([
    for (var i = 0; i < 16; i++)
      int.parse(hex.substring(i * 2, i * 2 + 2), radix: 16),
  ]);
}

String _formatUuid(Uint8List bytes) {
  final hex = [
    for (final byte in bytes) byte.toRadixString(16).padLeft(2, '0'),
  ];
  return [
    hex.sublist(0, 4).join(),
    hex.sublist(4, 6).join(),
    hex.sublist(6, 8).join(),
    hex.sublist(8, 10).join(),
    hex.sublist(10, 16).join(),
  ].join('-');
}
